### PR TITLE
Solver Refactor

### PR DESCRIPTION
**Pull request by mmcgurn**
This refactor is designed to prepare for solvers only using part of the domain. It:

- changes the flow so that a single dmPlex will be used by multiple solvers
- creates a subDomain based upon the plex label
- create a new unified solver interface
- moves the particles to the solver interface (same as flow solvers)

### CODE REVIEW

File: ablateLibrary/domain/subDomain.hpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace ablate::domain {
+
+class SubDomain {
+   private:
+    // const pointer to the parent domain/dm
+    const std::weak_ptr<Domain> domain;
+
+    // The label used to describe this subDomain
+    DMLabel label;
```
Comment (knepley) [inline ablateLibrary/domain/subDomain.hpp:19]: I think you might want (label, value) or even (label, {values}) to specify the subdomain

**Review by knepley — CHANGES_REQUESTED** [review]
Check out my comment about using the label. Otherwise it is great I think.

File: ablateLibrary/domain/subDomain.hpp (hunk 1 above)
Comment (knepley) [reply ablateLibrary/domain/subDomain.hpp:19]: Yep, that really needs to be fixed. I guess I should fix it now, before anyone else builds code on it.

Comment (mmcgurn) [reply ablateLibrary/domain/subDomain.hpp:22]: @knepley Thanks!  I've added a new struct for a region, that the solvers specify.  This includes a label name and values.  The subDomain now uses a label and values and defaults to "1".

**Pull request merged**